src/Menu/useMenu.ts

```diff
@@ -0,0 +1,281 @@
+            onSelect?.(event);
+         }
+      },
+      [onSelect, url]
+   );
+
+   const onMouseEnter = React.useCallback(() => {
+      focusContext.focus(index);
+   }, [focusContext.focus, itemRef, index]);
+
+   const onMouseMove = React.useCallback(() => {
+      if (document.activeElement !== itemRef.current) {
+         focusContext.focus(index);
+      }
+   }, [focusContext.focus, itemRef, index]);
+
+   const onMouseLeave = React.useCallback(() => {
+      focusContext.blur();
+   }, [focusContext.blur]);
+
+   const menuItemProps = React.useMemo((): ElementProps<any> &
+      MenuItemProps & { 'data-focus': boolean; href?: string } => {
+      return {
+         ...props,
+         href: url,
+         'data-focus': isFocused,
+         role: 'menuitem',
+         ref: itemRef,
+         tabIndex: isFocused ? 0 : -1,
+         onMouseEnter,
+         onMouseMove,
+         onMouseLeave,
+         onKeyDown,
+         onClick: onSelect,
+      };
+   }, [
+      props,
+      itemRef,
+      isFocused,
+      onMouseEnter,
+      onMouseMove,
+      onMouseLeave,
+      onSelect,
+      onKeyDown,
+      url,
+   ]);
+
+   return menuItemProps;
+}
+
+export type MenuButtonProps = React.HTMLAttributes<HTMLButtonElement>;
+
+export function useMenuButton(
+   props: MenuButtonProps
+): React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
+   const {
+      isOpen,
+      close,
+      openAndFocusFirstItem,
+      openAndFocusLastItem,
+      menuButtonRef,
+      menuId,
+      menuButtonId,
+   } = useMenuContext();
+
+   useUpdateEffect(() => {
+      if (!isOpen && menuButtonRef.current) {
+         menuButtonRef.current.focus();
+      }
+   }, [isOpen, menuButtonRef]);
+
+   const onClick = React.useCallback(() => {
+      if (isOpen) {
+         close();
+      } else {
+         openAndFocusFirstItem();
+      }
+   }, [isOpen, close, openAndFocusFirstItem]);
+
+   const onKeyDown = React.useCallback(
+      (event: React.KeyboardEvent) => {
+         const eventKey = normalizeEventKey(event);
+         const keyMap: { [key: string]: () => void } = {
+            Enter: openAndFocusFirstItem,
+            ArrowDown: openAndFocusFirstItem,
+            ArrowUp: openAndFocusLastItem,
+         };
+
+         const action = keyMap[eventKey];
+
+         if (action) {
+            event.preventDefault();
+            event.stopPropagation();
+            action();
+         }
+      },
+      [openAndFocusFirstItem, openAndFocusLastItem]
+   );
+
+   const buttonProps = React.useMemo((): React.DetailedHTMLProps<
+      React.ButtonHTMLAttributes<HTMLButtonElement>,
+      HTMLButtonElement
+   > => {
+      return {
+         ...props,
+         id: menuButtonId,
+         'aria-haspopup': 'menu',
+         'aria-expanded': isOpen,
+         'aria-controls': menuId,
+         onClick,
+         onKeyDown,
+         ref: menuButtonRef,
+      };
+   }, [isOpen, onClick, onKeyDown, props]);
```

Comment:
You know what? Now that you made me think about it, probably none of the `useMemo` that I used to return props bags are actually useful. The main reason I use `useMemo` is not to optimize the object creation, but to avoid unnecessary re-renders (due to callback or objects that get recreated on every render). In this case, those props would be just passed down to the components so we only need to make sure that the callbacks are memoized. I will remove all the unnecessary `useMemo` 👍